### PR TITLE
chore(release): 2.52.0 — n8n 2.20.x + community-node typeVersion fix (#781)

### LINKED ISSUE #781: Community nodes: get_node advertises npm package version as typeVersion, breaking workflows
## Summary

For community nodes, `get_node` returns the **npm package version** in the `version` field and emits a `versionNotice` telling the caller to use that value as `typeVersion`. The advertised value is often non-numeric (e.g. `0.2.21`) and almost never matches the node's actual registered `INodeTypeDescription.version` (typically `1` for declarative community nodes). Following the advice produces a workflow that n8n's runtime renders as a red/broken node.

## Repro

Tool call:
```
get_node({ nodeType: "nodes-base.playwright", detail: "standard" })
```

Response (excerpt) for `n8n-nodes-playwright` (npm 0.2.21):
```json
{
  "version": "0.2.21",
  "isVersioned": false,
  "versionNotice": "⚠️ Use typeVersion: 0.2.21 when creating this node",
  "metadata": { "developmentStyle": "declarative" }
}
```

Then:
```
n8n_create_workflow({ nodes: [{ type: "n8n-nodes-playwright.playwright", typeVersion: 0.2, ... }] })
validate_workflow(...)   // ✅ passes
n8n_validate_workflow(id) // ✅ passes
```

In the n8n UI: the node renders as **red/broken** even though `n8n-nodes-playwright` is installed and listed in *Settings → Community Nodes*. Manually re-adding the node from the UI fixes it (the UI uses the runtime-registered version, presumably `1`).

## Why this is a bug

1. **`0.2.21` isn't a valid JS number** — it has two dots. There is no way to assign it literally to `typeVersion`, which is typed as `number`. Truncating to `0.2` makes it numeric but doesn't match anything the runtime knows about.
2. **`version` looks like it's sourced from `package.json#version`** (or the npm metadata cache), not from the node's `INodeTypeDescription.version`. For community nodes these two values are unrelated — the package version moves with releases, the node's typeVersion is a contract with workflow JSON.
3. **`isVersioned: false` is already set** — so the system knows there's no real version history — yet `versionNotice` still emits a confident-sounding directive.
4. **Validators don't catch it.** Both `validate_workflow` and `n8n_validate_workflow` accept `typeVersion: 0.2` because they compare against the same (incorrect) advertised value rather than the runtime-registered descriptor.

## Suggested fixes

In rough priority order:

- [ ] For community nodes, source `version` from the node descriptor (`INodeTypeDescription.version`) — not `package.json#version`. If the descriptor isn't ingestible, default to `1` (which is what the vast majority of declarative community nodes register).
- [ ] Suppress or rewrite `versionNotice` when `isVersioned: false`, or when the resolved `version` isn't a finite JS number. Today's notice is actively misleading.
- [ ] In `validate_workflow` / `validate_node`, reject `typeVersion` values that aren't finite numbers (e.g. `0.2.21`, `NaN`, strings) before any further checks.
- [ ] Optionally: add a 'community node' note to the `versionNotice` itself explaining that community node typeVersion is whatever the node's descriptor declares (often `1`), independent of the npm release.

## Environment

- n8n-mcp tool path used: `mcp__claude_ai_n8n-mcp-production__*` (Claude.ai-hosted production MCP)
- Affected community node: `n8n-nodes-playwright` v0.2.21 (author: toema)
- Confirmed by manually re-adding the node in n8n's UI (works) vs. inserting via `n8n_create_workflow` with `typeVersion: 0.2` (broken)

## Related context

Discovered while building a sample headless-screenshot workflow with `n8n-nodes-playwright`. The end-to-end symptom for the user was "node shows as red/broken in the UI even though the package is installed" — which is hard to debug because both validators report the workflow as valid.

### PR DESCRIPTION
**Pull request by czlonkowski**
## Summary

Two changes bundled in one release:

### 1. Updated n8n dependencies to 2.20.x

| Package | From | To |
|---|---|---|
| `n8n-nodes-base` | 2.18.3 | 2.20.4 |
| `n8n-core` | 2.18.3 | 2.20.3 |
| `n8n-workflow` | 2.18.3 | 2.20.0 |
| `@n8n/n8n-nodes-langchain` | 2.18.3 | 2.20.4 |

Database rebuilt against the new packages; the 830 community node rows were preserved across the rebuild.

### 2. Fix #781 — community-node `typeVersion` correctness

For community nodes, `get_node` previously returned the **npm package version** (e.g. `"0.2.21"`) in the `version` field and emitted:

```
versionNotice: "⚠️ Use typeVersion: 0.2.21 when creating this node"
```

`0.2.21` is not a valid JS number — assigning `typeVersion: 0.2` produced workflows that n8n's runtime rendered as **red/broken nodes** even though both `validate_workflow` and `n8n_validate_workflow` reported them as valid. Manually re-adding the node from the n8n UI fixed it (the UI uses the runtime-registered version, typically `1`).

#### Root cause
- `community-node-service.ts` seeded the `version` column from `attributes.npmVersion` (Strapi path) or `pkgInfo.version` (npm-only path) when the descriptor's version was missing — conflating npm package version with `INodeTypeDescription.version`.
- `getNodeEssentials` echoed that string back into `versionNotice` with no numeric coercion.
- `WorkflowValidator.validateAllNodes` compared `node.typeVersion < nodeInfo.version` directly, with `nodeInfo.version` being a string. JS coerces `"0.2.21"` to `NaN`, so neither the "outdated" warning nor the "exceeds maximum" error fired — the bogus typeVersion silently passed.

#### Three-part fix
- **Parser** (`src/community/community-node-service.ts`): drop the `npmVersion` fallback for Strapi nodes and stop seeding the npm version for npm-only nodes. Both default to `'1'` (what declarative community nodes register at runtime). The npm package version is preserved separately in `npm_version`.
- **`get_node` response** (`src/mcp/server.ts`): coerces `version` through a new `parseTypeVersion` helper. For community nodes the response now also includes `isCommunity: true`, `npmVersion`, a community-aware `versionNotice`, and a `metadata.versionCoerced` audit field whenever stale seed data has to be resolved on the fly (so callers and downstream tooling can see when defensive coercion happened).
- **Validator** (`src/services/workflow-validator.ts`): rejects non-finite typeVersions (including `NaN`) with an explicit *"must be a finite non-negative number"* message, parses comma-separated and array-form `nodeInfo.version` strings before min/max comparisons, falls back to suggesting `typeVersion: 1` when the database version is unparseable, and stops treating literal `0` as a missing typeVersion.

#### Data migration in the shipped DB
`data/nodes.db`: 118 community rows whose stored `version` was a multi-dot semver or contained letters were reset to `'1'` in place — so the fix is visible to anyone pulling this release without waiting for the next community-node resync.

#### New shared helper
`src/utils/typeversion.ts` — `parseTypeVersion()` and `isValidTypeVersion()` are the single source of truth across parser, get_node response, and validator. Handles numbers, number arrays, comma-separated strings, JSON arrays, and rejects npm-style multi-dot semver strings.

## Test plan

- [x] `npm run typecheck` — clean
- [x] `npm run build` — clean
- [x] `npx vitest run tests/unit` — **4772 / 4772 pass** (35 skipped, no new failures)
- [x] New tests:
  - `tests/unit/utils/typeversion.test.ts` (17 cases — numbers, arrays, semver rejection, JSON, edge cases)
  - `tests/unit/community/community-node-service.test.ts` — Strapi/npm parser fix, descriptor arrays, npm-version isolation
  - `tests/unit/services/workflow-validator.test.ts` — NaN rejection, unparseable `nodeInfo.version` handling, comma-separated max, fallback suggestion
- [x] DB sanity: `n8n-nodes-playwright.playwright` row now has `version=1, npm_version=0.2.21, is_community=1`
- [ ] Manual: reload the MCP server and confirm `get_node({ nodeType: "nodes-base.playwright", detail: "standard" })` returns `version: 1`, `isCommunity: true`, `npmVersion: "0.2.21"`, and the community-aware `versionNotice`
- [ ] Manual: confirm a workflow with `typeVersion: 1` for `n8n-nodes-playwright.playwright` renders correctly in the n8n UI

## Notes

- `getNodeInfo` (full-mode `get_node`) still returns the raw stored `version` from the DB. It does not synthesize a `versionNotice`, so it isn't affected by the misleading-advice problem the issue describes. Anyone relying on it gets the data fix from the in-place `data/nodes.db` migration.
- Reporter: @czlonkowski.

Conceived by Romuald Członkowski - https://www.aiadvisors.pl/en

### CODE REVIEW

File: src/mcp/server.ts [hunk 1]
```diff
@@ -2843,8 +2853,7 @@ Full documentation is being prepared. For now, use get_node_essentials for confi
       category: node.category,
       version: latestVersion,
       isVersioned: node.isVersioned ?? false,
-      // Prominent warning to use the correct typeVersion
-      versionNotice: `⚠️ Use typeVersion: ${latestVersion} when creating this node`,
+      versionNotice,
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/mcp/server.ts:2856]: Documented as a Changed entry in e5e3d5e. The number form is the correct shape for typeVersion (assignable straight into workflow JSON without coercion); reverting only for non-community nodes would mean the same field has two different types depending on which row you look at, which is worse than the explicit one-time contract change.

File: src/utils/typeversion.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    // Reject npm-package-style multi-dot strings like "0.2.21" or "2.1.17-rc.31".
+    if ((trimmed.match(/\./g) || []).length > 1) return null;
+
+    const n = Number(trimmed);
+    return Number.isFinite(n) ? n : null;
+  }
+
+  return null;
+}
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/utils/typeversion.ts:58]: Fixed in e5e3d5e — `parseTypeVersion` now rejects negatives (number, numeric string, and `-Infinity`) so it agrees with `isValidTypeVersion` and the validator's `< 0` check. typeVersion 0 remains valid. New unit cases for `-1`, `-0.5`, `'-1'`, `'-0.5'`.

File: src/utils/typeversion.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    if (trimmed.includes(',')) {
+      return parseTypeVersion(trimmed.split(',').map((s) => s.trim()));
+    }
+
+    // Reject npm-package-style multi-dot strings like "0.2.21" or "2.1.17-rc.31".
+    if ((trimmed.match(/\./g) || []).length > 1) return null;
+
+    const n = Number(trimmed);
+    return Number.isFinite(n) ? n : null;
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/utils/typeversion.ts]: Pushing back on this one. `INodeTypeDescription.version` is typed as `number | number[]` in n8n-workflow — legitimate descriptors don't emit "1.2.0" strings. Multi-dot strings only appear when something has stuffed a semver into the field, which is exactly the npm-package-version footgun #781 fixes. Parsing only the first two components ("1.2.0" → 1.2) would be guessing at intent and could mask the same class of bug for a node whose actual descriptor version is 1 — the misread would be silent. Falling back to 1 (the default for declarative community nodes) plus the new `metadata.versionCoerced` audit field on the `get_node` response is intentionally noisy when seed data is bad.

File: src/services/workflow-validator.ts [hunk 4]
```diff
@@ -496,40 +497,48 @@ export class WorkflowValidator {
...
-          else if (nodeInfo.version && node.typeVersion > nodeInfo.version) {
+          else if (maxVersion !== null && node.typeVersion > maxVersion) {
             result.errors.push({
               type: 'error',
               nodeId: node.id,
               nodeName: node.name,
-              message: `typeVersion ${node.typeVersion} exceeds maximum supported version ${nodeInfo.version}`
+              message: `typeVersion ${node.typeVersion} exceeds maximum supported version ${maxVersion}`
             });
           }
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/services/workflow-validator.ts:554]: Fixed in e5e3d5e — when `nodeInfo.isVersioned` is true and `maxVersion` comes back null, the validator now emits a `Cannot validate typeVersion for <type>: stored version "<value>" is not a valid typeVersion. Min/max checks were skipped` warning. New test covers the exact scenario you flagged: `typeVersion: 999` against a node with stored version `"0.2.21"` no longer silently passes — the warning surfaces the gap so callers know the upper bound wasn't enforced.

File: src/community/community-node-service.ts [hunk 5]
```diff
@@ -280,7 +281,9 @@ export class CommunityNodeService {
         nodeDesc.group?.includes('webhook') ||
         false,
       isVersioned: (attributes.nodeVersions?.length || 0) > 1,
-      version: nodeDesc.version?.toString() || attributes.npmVersion || '1',
+      // typeVersion is the descriptor's version, NOT the npm package version.
+      // npm version (e.g. "0.2.21") is exposed separately via npmVersion below.
+      version: (parseTypeVersion(nodeDesc.version) ?? 1).toString(),
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply src/community/community-node-service.ts:286]: Acknowledged — the collapse to max is intentional for community nodes specifically. Two reasons:

1. Community nodes don't populate the `node_versions` table (`getNodeVersions` and friends in node-repository.ts only return rows for base nodes, populated during the npm-package rebuild step). So the comma-separated form was never enumerated downstream for community rows — it was just a string the validator's old code happened to mis-compare against.

2. For the `get_node` response and the validator's max-check, only the latest typeVersion matters. The validator handles both forms (single number and comma-separated arrays) since base nodes still produce the comma form. Community-node arrays collapsing to max keeps the response cleaner and avoids the impression that a community node "supports" arbitrary historical typeVersions when the descriptor is silent on that.

Documented as the explicit test `'Strapi: collapses descriptor version arrays to the highest entry'` in the community-node-service suite.

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+Conceived by Romuald Członkowski - https://www.aiadvisors.pl/en
+
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply CHANGELOG.md:22]: The empty `## [Unreleased]` above the most recent dated section is the existing pattern in this file — see 2.51.3 (2026-05-11), 2.51.2 (2026-05-11), 2.51.1 (2026-05-06), and earlier entries; the unreleased header is kept and the new dated section is inserted immediately below it on each release. This PR follows that convention. Re: the hard-coded date — the merge is happening within the same day window, so the date will be accurate; will update if it slips.

File: package.json [hunk 7]
```diff
@@ -156,16 +156,16 @@
...
     "express": "^5.1.0",
     "express-rate-limit": "^7.1.5",
     "form-data": "^4.0.5",
     "lru-cache": "^11.2.1",
-    "n8n-core": "2.18.3",
-    "n8n-nodes-base": "2.18.3",
-    "n8n-workflow": "2.18.3",
+    "n8n-core": "^2.20.3",
+    "n8n-nodes-base": "^2.20.4",
+    "n8n-workflow": "^2.20.0",
```
[the remark this answers is not in the document]
Comment (czlonkowski) [reply package.json]: Fixed in e5e3d5e — re-pinned to exact versions (`2.20.4`, `2.20.3`, `2.20.0`, `2.20.4`) and patched `scripts/update-n8n-deps.js` so future updates write exact pins for these four packages instead of `^${version}`. The script always wrote the caret form, which is how it crept in here — and exactly the regression you flagged: with `^2.20.4`, the lock regen pulled `2.21.1` actual install, which is a different node set than `data/nodes.db` was rebuilt against. With the exact pin, lock now resolves to `2.20.4` for real (`require('n8n-nodes-base/package.json').version` returns `'2.20.4'` after a clean install).
